test/unit/sys-execute.c

```diff
@@ -0,0 +1,167 @@
+
+#include "tss2_sys.h"
+#include "sysapi_util.h"
+#include "tss2-tcti/tcti-common.h"
+
+
+#define LOGMODULE test
+#include "util/log.h"
+
+/**
+ * Test calls Tss2_Sys_Execute() many times after receiving TPM2_RC_RETRY
+ */
+
+const uint8_t ok_response[] = {
+    0x80, 0x01,                 /* TPM_ST_NO_SESSION */
+    0x00, 0x00, 0x00, 0x2C,     /* Response Size 10 + 2 + 32 */
+    0x00, 0x00, 0x00, 0x00,     /* TPM_RC_SUCCESS */
+    0x00, 0x20,                 /* size of buffer */
+    0xde, 0xad, 0xbe, 0xef,
+    0xde, 0xad, 0xbe, 0xef,
+    0xde, 0xad, 0xbe, 0xef,
+    0xde, 0xad, 0xbe, 0xef,
+    0xde, 0xad, 0xbe, 0xef,
+    0xde, 0xad, 0xbe, 0xef,
+    0xde, 0xad, 0xbe, 0xef,
+    0xde, 0xad, 0xbe, 0xef,
+};
+
+const uint8_t retry_response[] = {
+    0x80, 0x01,                 /* TPM_ST_NO_SESSION */
+    0x00, 0x00, 0x00, 0x0A,     /* Response Size 10 */
+    0x00, 0x00, 0x09, 0x22      /* TPM2_RC_RETRY */
+};
+TSS2_TCTI_COMMON_CONTEXT _tcti_ctx;
+TSS2_TCTI_CONTEXT_COMMON_V1 *_tcti_v1_ctx;
+
+TSS2_RC
+tcti_transmit(
+    TSS2_TCTI_CONTEXT *tctiContext,
+    size_t size,
+    uint8_t const *command)
+{
+    TSS2_RC r = 0;
+    tpm_header_t hdr;
+
+    LOG_DEBUG ("%s: transmiting %zu bytes", __func__, size);
+    r = header_unmarshal (command, &hdr);
+    if (r)
+        return r;
+
+    LOG_DEBUG ("%s request_hdr.tag  = %x", __func__, hdr.tag);
+    LOG_DEBUG ("%s request_hdr.size  = %x", __func__, hdr.size);
+    LOG_DEBUG ("%s request_hdr.code  = %x", __func__, hdr.code);
+
+    if (hdr.tag != TPM2_ST_NO_SESSIONS || hdr.size != 0xC || hdr.code != 0x17B)
+        return TSS2_TCTI_RC_BAD_VALUE;
+
+    return r;
+}
+
+#define NUM_OF_RETRIES 4
+
+TSS2_RC
+tcti_receive(
+    TSS2_TCTI_CONTEXT *tctiContext,
+    size_t *size,
+    uint8_t *response,
+    int32_t timeout)
+{
+    static int i;
+
+    LOG_DEBUG ("%s: receiving response, size %zu, buff %p",
+               __func__, sizeof(ok_response), response);
+
+    if (response == NULL) {
+        *size = sizeof(ok_response);
+        return TPM2_RC_SUCCESS;
+    }
+
+    if (i++ < NUM_OF_RETRIES) {
+        LOG_DEBUG ("%s: return RC_RETRY", __func__);
+        memcpy(response, retry_response, sizeof(retry_response));
+        *size = sizeof(retry_response);
+        return TPM2_RC_SUCCESS;
+    }
+    memcpy(response, ok_response, sizeof(ok_response));
+    *size = sizeof(ok_response);
+    return TPM2_RC_SUCCESS;
+}
+
+TSS2_ABI_VERSION ver = TSS2_ABI_VERSION_CURRENT;
+
+static int
+setup(void **state)
+{
+    TSS2_SYS_CONTEXT  *sys_ctx;
+    TSS2_TCTI_CONTEXT *tcti_ctx = (TSS2_TCTI_CONTEXT *) &_tcti_ctx;
+    UINT32 size_ctx;
+    TSS2_RC r;
+
+    size_ctx = Tss2_Sys_GetContextSize(0);
+    sys_ctx = calloc (1, size_ctx);
+    assert_non_null (sys_ctx);
+    _tcti_v1_ctx = (TSS2_TCTI_CONTEXT_COMMON_V1 *) &_tcti_ctx;
+    _tcti_v1_ctx->version = 1;
+    _tcti_v1_ctx->transmit = tcti_transmit;
+    _tcti_v1_ctx->receive = tcti_receive;
+    _tcti_ctx.state = TCTI_STATE_TRANSMIT;
```

Comment:
This variable is never used.

Reply:
Do you mean _tcti_ctx? We need to set the internal TCTI state to transmit otherwise the Execute() will fail with TSS2_TCTI_RC_BAD_SEQUENCE https://github.com/tpm2-software/tpm2-tss/blob/master/src/tss2-tcti/tcti-common.c#L50
Also this is the storage space for the whole TCTI context.
Both the tcti_ctx * and _tcti_v1_ctx * point to &_tcti_ctx, which is TSS2_TCTI_COMMON_CONTEXT.

Reply:
`.state` is never used. And tcti_common_transmit_checks is not called from tcti_transmit function in this file.

I guess none of the extra fields from TSS2_TCTI_COMMON_CONTEXT are called, so using TSS2_TCTI_CONTEXT_COMMON_V1 would even be sufficient for _tcti_ctx...

Reply:
Ok, you are right, TSS2_TCTI_CONTEXT_COMMON_V1 is sufficient to setup the test. Code updated.